cli/plasmo/src/type.ts

```diff
@@ -5,50 +5,103 @@ export type PlasmoContentScript = Omit<Partial<ManifestContentScript>, "js">
 
 type Async<T> = Promise<T> | T
 
-type Getter<T> = () => Async<T>
+type Getter<T, P = any> = (props?: P) => Async<T>
 
-type GetHtmlElement = Getter<HTMLElement>
+type GetElement = Getter<Element>
 
-export type PlasmoGetRootContainer = GetHtmlElement
-export type PlasmoGetOverlayAnchor = GetHtmlElement
+export type PlasmoCSUIAnchor = {
+  element: Element
+  type: "overlay" | "inline"
+}
 
-export type PlasmoGetInlineAnchor = () => HTMLElement | null
+export type PlasmoCSUIProps = {
+  anchor?: PlasmoCSUIAnchor
+}
 
 export type PlasmoCSUIMountState = {
   document: Document
   observer: MutationObserver | null
-  shadowHost: Element | null
-  inlineAnchor: Element | null
+
+  isMounting: boolean
+  isMutated: boolean
+  /**
+   * Used to quickly check if element is already mounted
+   */
+  hostSet: Set<Element>
+
+  /**
+   * Used to add more metadata to the host Set
+   */
+  hostMap: WeakMap<Element, PlasmoCSUIAnchor>
+
+  /**
+   * Used to align overlay anchor with elements on the page
+   */
+  overlayTargetList: Element[]
 }
 
-export type PlasmoMountShadowHost = (
-  mountState: PlasmoCSUIMountState
-) => Async<void>
+export type PlasmoGetRootContainer = (
+  props: {
+    mountState?: PlasmoCSUIMountState
+  } & PlasmoCSUIProps
+) => Async<Element>
 
-export type PlasmoRender = (
-  createRootContainer: GetHtmlElement,
-  MountContainer: () => JSX.Element | HTMLElement
+export type PlasmoGetOverlayAnchor = GetElement
+export type PlasmoGetOverlayAnchorList = Getter<NodeList>
+
+export type PlasmoGetInlineAnchor = GetElement
+export type PlasmoGetInlineAnchorList = Getter<NodeList>
+
+export type PlasmoMountShadowHost = (
+  props: {
+    observer: MutationObserver | null
+    shadowHost: Element
+  } & PlasmoCSUIProps
 ) => Async<void>
 
-export type PlasmoGetShadowHostId = Getter<string>
+export type PlasmoGetShadowHostId = Getter<string, PlasmoCSUIAnchor>
 
-export type PlasmoGetStyle = Getter<HTMLStyleElement>
+export type PlasmoGetStyle = Getter<HTMLStyleElement, PlasmoCSUIAnchor>
 
+/**
+ * @return a cleanup unwatch function that will be run when unmounted
+ */
 export type PlasmoWatchOverlayAnchor = (
   updatePosition: () => Promise<void>
-) => void
+) => () => void
+
+export type PlasmoCSUIContainerProps = {
+  id?: string
+  children?: React.ReactNode
+  watchOverlayAnchor?: PlasmoWatchOverlayAnchor
+} & PlasmoCSUIProps
+
+export type PlasmoCSUIContainer = (
+  p: PlasmoCSUIContainerProps
+) => JSX.Element | Element
 
 export type PlasmoCreateShadowRoot = (
-  shadowHost: HTMLDivElement
+  shadowHost: HTMLElement
 ) => Async<ShadowRoot>
 
+export type PlasmoRender = (
+  props: {
+    createRootContainer?: (p: PlasmoCSUIAnchor) => Async<Element>
```

Comment:
Does the first parameter need to be required?  I was wondering how we'd use this if we're not mounting into multiple elements, like in the following example.

https://github.com/PlasmoHQ/examples/blob/b3f699d537c5143c928ede46586901bff1cfca62/with-content-scripts-ui/contents/plasmo-root-container.tsx